File: LC 882.cpp
using pii = pair<int, int>;
class Solution {
    vector<int> dijkstra(vector<vector<pii>>& g, int start) {
        int n = g.size();
        vector<int> dist(n, 0x7f7f7f7f), visited(n, 0);
        priority_queue<pii, vector<pii>, greater<pii>> pq;
        pq.push(make_pair(0, start));
        dist[start] = 0;
        while (!pq.empty()) {
            int u = pq.top().second;
            pq.pop();
            if (visited[u]) continue;
            visited[u] = 1;
            for (auto& edge : g[u]) {
                int v= edge.first, weight = edge.second;
                if (!visited[v] && dist[u] + weight < dist[v]) {
                    dist[v] = dist[u] + weight;
                    pq.push(make_pair(dist[v], v));
                }
            }
        }
        return dist;
    }
public:
    int reachableNodes(vector<vector<int>>& edges, int maxMoves, int n) {
        vector<vector<pii>> g(n);
        for (auto& edge : edges) {
            int a = edge[0], b = edge[1], c = edge[2];
            g[a].push_back(make_pair(b, c + 1));
            g[b].push_back(make_pair(a, c + 1));
        }
        auto dist = dijkstra(g, 0);
        int ans = 0;
        for (int x : dist) {
            if (x <= maxMoves)
                ++ans;
        }
        for (auto& e : edges) {
            int a = e[0], b = e[1], c = e[2];
            int x = max(0, maxMoves - dist[a]);
            int y = max(0, maxMoves - dist[b]);
            ans += min(x + y, c);
        }
        return ans;
    }
};

File: 1000/LC 882.go
func reachableNodes(edges [][]int, maxMoves int, n int) (ans int) {
    g := make([][]neighbor, n)
    for _, e := range edges {
        u, v, cnt := e[0], e[1], e[2]
        g[u] = append(g[u], neighbor{v, cnt + 1})
        g[v] = append(g[v], neighbor{u, cnt + 1})
    }

    dist := dijkstra(g, 0)

    for _, d := range dist {
        if d <= maxMoves {
            ans++
        }
    }

    for _, e := range edges {
        u, v, cnt := e[0], e[1], e[2]
        a := max(maxMoves - dist[u], 0)
        b := max(maxMoves - dist[v], 0)
        ans += min(a + b, cnt)
    }
    return
}

type neighbor struct {
    to, weight int
}

func dijkstra(g [][]neighbor, start int) []int {
    dist := make([]int, len(g))
    for i := range dist {
        dist[i] = math.MaxInt
    }
    dist[start] = 0
    h := hp{{start, 0}}
    for h.Len() > 0 {
        t := heap.Pop(&h).(pair)
        x := t.x
        if dist[x] < t.dist {
            continue
        }
        for _, e := range g[x] {
            y := e.to
            if d := dist[x] + e.weight; d < dist[y] {
                dist[y] = d
                heap.Push(&h, pair{y, d})
            }
        }
    }

    return dist
}

type pair struct {
    x, dist int
}
type hp []pair
func (h hp) Len() int {return len(h)}
func (h hp) Less(i, j int) bool {return h[i].dist < h[j].dist}
func (h hp) Swap(i, j int) {h[i], h[j] = h[j], h[i]}
func (h *hp) Push(x any) {
    *h = append(*h, x.(pair))
}
func (h *hp) Pop() (x any) {
    old := *h
    n := len(old)
    x, *h = old[n-1], old[:n-1]
    return
}